monkey/tests/unit_tests/monkey_island/cc/agent_event_handlers/test_handle_ping_scan_event.py

```diff
@@ -0,0 +1,239 @@
+)
+
+EVENT_NO_RESPONSE = PingScanEvent(
+    source=AGENT_ID,
+    target=IPv4Address("10.10.10.1"),
+    response_received=False,
+    os=OperatingSystem.LINUX,
+)
+
+
+@pytest.fixture
+def agent_repository() -> IAgentRepository:
+    agent_repository = MagicMock(spec=IAgentRepository)
+    agent_repository.upsert_agent = MagicMock()
+    agent_repository.get_agent_by_id = MagicMock(return_value=AGENT)
+    return agent_repository
+
+
+@pytest.fixture
+def machine_repository() -> IMachineRepository:
+    machine_repository = MagicMock(spec=IMachineRepository)
+    machine_repository.get_new_id = MagicMock(side_effect=count(SEED_ID))
+    machine_repository.upsert_machine = MagicMock()
+    return machine_repository
+
+
+@pytest.fixture
+def node_repository() -> INodeRepository:
+    node_repository = MagicMock(spec=INodeRepository)
+    node_repository.upsert_communication = MagicMock()
+    return node_repository
+
+
+@pytest.fixture
+def handler(agent_repository, machine_repository, node_repository) -> handle_ping_scan_event:
+    return handle_ping_scan_event(agent_repository, machine_repository, node_repository)
+
+
+machines = {MACHINE_ID: PINGER_MACHINE, TARGET_MACHINE.id: TARGET_MACHINE}
+machines_by_id = {MACHINE_ID: PINGER_MACHINE, TARGET_MACHINE.id: TARGET_MACHINE}
+machines_by_ip = {
+    IPv4Address("10.10.10.99"): [PINGER_MACHINE],
+    IPv4Address("10.10.10.1"): [TARGET_MACHINE],
+}
+
+
+def machine_from_id(id: int):
+    return machines_by_id[id]
+
+
+def machines_from_ip(ip: IPv4Address):
+    return machines_by_ip[ip]
+
+
+class error_machine_by_id:
+    """Raise an error if the machine with the called ID matches the stored ID"""
+
+    def __init__(self, id: int, error):
+        self.id = id
+        self.error = error
+
+    def __call__(self, id: int):
+        if id == self.id:
+            raise self.error
+        else:
+            return machine_from_id(id)
+
+
+class error_machine_by_ip:
+    """Raise an error if the machine with the called IP matches the stored ID"""
+
+    def __init__(self, id: int, error):
+        self.id = id
+        self.error = error
+
+    def __call__(self, ip: IPv4Address):
+        print(f"IP is: {ip}")
+        machines = machines_from_ip(ip)
+        if machines[0].id == self.id:
+            print(f"Raise error: {self.error}")
+            raise self.error
+        else:
+            print(f"Return machine: {machines}")
+            return machines
+
+
+def test_handle_ping_scan_event__upserts_machine(
+    handler: handle_ping_scan_event,
+    machine_repository: IMachineRepository,
+):
+    machine_repository.get_machine_by_id = MagicMock(side_effect=machine_from_id)
+    machine_repository.get_machines_by_ip = MagicMock(side_effect=UnknownRecordError)
+    handler(EVENT)
+
+    expected_machine = TARGET_MACHINE.copy()
+    expected_machine.operating_system = OperatingSystem.LINUX
+
+    machine_repository.upsert_machine.assert_called_with(expected_machine)
```

Comment:
Try to make tests follow the Arrange, Act, Assert (AAA) pattern

```suggestion
    # Arrange
    machine_repository.get_machine_by_id = MagicMock(side_effect=machine_from_id)
    machine_repository.get_machines_by_ip = MagicMock(side_effect=UnknownRecordError)
    expected_machine = TARGET_MACHINE.copy()
    expected_machine.operating_system = OperatingSystem.LINUX
    
    # Act
    handler(EVENT)
    
    # Assert
    machine_repository.upsert_machine.assert_called_with(expected_machine)
```

Reply:
I figured it should go in the Assert step because it is only used in the Assert step -- it is not need for the Act step. That fits with the practice of defining variables as close as possible to where they're used. However, I'll go with whatever policy the team uses

Reply:
In my brain, defining expected outputs is part of Arranging. I can see the case for Assert.

The original version of this had Act and Arrange together, than something with the expected value, then Assert.

I'm fine either way, as long as the Act is clearly separated from Arrange and Assert.